sklearn/preprocessing/_target_encoder.py

```diff
@@ -311,30 +360,114 @@ def fit_transform(self, X, y):
+        """Precompute all fast-path structures during fit/fit_transform."""
+
+        n_features = len(self.categories_)
+        is_multi = self.target_type_ == "multiclass"
+        n_classes = 0 if not is_multi else len(self.classes_)
+
+        self._te_is_multiclass_ = is_multi
+        self._te_index_maps_ = [None] * n_features
+        self._te_enc_vecs_ = [None] * n_features
+        self._te_enc_blocks_ = [None] * n_features
+        self._te_defaults_ = [None] * n_features
+
+        for j in range(n_features):
+            cats_j = self.categories_[j]
+            idx_map = {}
+            for i, cat in enumerate(cats_j):
+                k = _norm_key(cat)
+                idx_map[k] = i
+            self._te_index_maps_[j] = idx_map
+            if not is_multi:
+                enc_vec = np.asarray(self.encodings_[j], dtype=float)
+                self._te_enc_vecs_[j] = enc_vec
+                self._te_defaults_[j] = float(np.asarray(self.target_mean_))
+            else:
+                block = np.empty((n_classes, len(cats_j)), dtype=float)
+                base = j * n_classes
+                for c in range(n_classes):
+                    block[c, :] = self.encodings_[base + c]
+                self._te_enc_blocks_[j] = block
+
+                default = np.asarray(self.target_mean_, dtype=float)
+                self._te_defaults_[j] = default
+
     def transform(self, X):
         """Transform X with the target encoding.
 
-        .. note::
-            `fit(X, y).transform(X)` does not equal `fit_transform(X, y)` because a
-            :term:`cross fitting` scheme is used in `fit_transform` for encoding.
-            See the :ref:`User Guide <target_encoder>`. for details.
-
         Parameters
         ----------
         X : array-like of shape (n_samples, n_features)
-            The data to determine the categories of each feature.
+            Input data to encode. Missing values (e.g. ``None`` or ``np.nan``)
+            are treated as categories. Categories unseen during :meth:`fit`
+            are encoded with ``target_mean_``.
 
         Returns
         -------
         X_trans : ndarray of shape (n_samples, n_features) or \
-                    (n_samples, (n_features * n_classes))
-            Transformed input.
+                (n_samples, n_features * n_classes)
+            Encoded representation of ``X``. For binary and continuous targets,
+            one column per input feature is returned. For multiclass targets,
+            one column per (feature, class) pair is returned, with classes ordered
+            as in ``classes_``.
         """
+        # Use the fast path for small batch else fall back to the vectorized path.
+        check_is_fitted(self)
+        X_checked = validate_data(
+            self,
+            X,
+            reset=False,
+            ensure_2d=True,
+            dtype=None,
+            ensure_all_finite="allow-nan",
+        )
+        n_samples = X_checked.shape[0]
```

Comment:
if we just want the number of samples, we don't have to call validation on X, we could just call `_n_samples` on it